### PR TITLE
[wasm] Vector128's Sum and Dot SIMD implementation

### PR DESCRIPTION
**Pull request by radekdoulik**
Implements `Vector128.Sum` and `Vector128.Dot` SIMD intrinsics on WebAssembly. It calculates the sum by shuffling (with swizzle instruction) the lanes around and doing vector addition. This way it takes log2(number of lanes) additions.

An example of emitted code:
```
(func Wasm_Browser_Bench_Sample_Sample_VectorTask_DotFloat_RunStep(param $0 i32, $1 i32))
 local $2 v128
 local.get $0
 i32.eqz
 if
  call mini_llvmonly_throw_nullref_exception
  unreachable

 local.get $0
 local.get $0
 v128.load offset:16    [SIMD]
 local.get $0
 v128.load offset:32    [SIMD]
 f32x4.mul    [SIMD]
 local.tee $2
 local.get $2
 v128.const 0x0b0a09080f0e0d0c0302010007060504    [SIMD]
 i8x16.swizzle    [SIMD]
 f32x4.add    [SIMD]
 local.tee $2
 local.get $2
 v128.const 0x07060504030201000f0e0d0c0b0a0908    [SIMD]
 i8x16.swizzle    [SIMD]
 f32x4.add    [SIMD]
 f32x4.extract.lane 0    [SIMD]
 f32.store offset:48 align:2
```
C# method:
```
public override void RunStep() {
    result = Vector128.Dot(vector1, vector2);
}
```

bench sample measurements on amd64 desktop:

Chrome/amd64

| measurement | before | after | times faster |
|-:|-:|-:|-:|
|                Vector, Dot product int |     0.4398us |     0.0463us |  9.5 |
|              Vector, Dot product ulong |     0.2567us |     0.0474us |  5.4 |
|              Vector, Dot product float |     0.4305us |     0.0482us |  8.9 |
|             Vector, Dot product double |     0.2519us |     0.0458us |  5.5 |
|                      Vector, Sum sbyte |     0.8576us |     0.0463us | 18.5 |
|                      Vector, Sum short |     0.4632us |     0.0457us | 10.1 |
|                       Vector, Sum uint |     0.2686us |     0.0454us |  5.9 |
|                     Vector, Sum double |     0.1633us |     0.0480us |  3.4 |

Firefox/amd64

| measurement | before | after | times faster |
|-:|-:|-:|-:|
|                Vector, Dot product int |     0.3993us |     0.1108us | 3.6 |
|              Vector, Dot product ulong |     0.2668us |     0.1081us | 2.5 |
|              Vector, Dot product float |     0.3897us |     0.1032us | 3.8 |
|             Vector, Dot product double |     0.2592us |     0.1096us | 2.4 |
|                      Vector, Sum sbyte |     0.6510us |     0.1047us | 6.2 |
|                      Vector, Sum short |     0.4038us |     0.1025us | 3.9 |
|                       Vector, Sum uint |     0.2631us |     0.1025us | 2.6 |
|                     Vector, Sum double |     0.1928us |     0.1022us | 1.9 |

### CODE REVIEW

**Comment by radekdoulik** [comment]
One possible future improvement for signed 16bit integers will be to use `i32x4.dot_i16x8_s(a: v128, b: v128) -> v128` instruction.

**Comment by radical** [comment]
Is there a test for this in `src/tests`?

File: src/mono/mono/mini/simd-intrinsics.c [hunk 1]
```diff
@@ -1276,14 +1281,17 @@ emit_sri_vector (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsi
...
 		if (!is_element_type_primitive (fsig->params [0]))
 			return NULL;
 
 		int instc0 = type_enum_is_float (arg0_type) ? OP_FMUL : OP_IMUL;
 		MonoInst *pairwise_multiply = emit_simd_ins_for_sig (cfg, klass, OP_XBINOP, instc0, arg0_type, fsig, args);
-
+#if defined(TARGET_ARM64)
 		return emit_sum_vector (cfg, fsig->params [0], arg0_type, pairwise_multiply);
+#else
+		return emit_sum_vector (cfg, klass, fsig, arg0_type, &pairwise_multiply);
```
Comment (vargaz) [inline src/mono/mono/mini/simd-intrinsics.c:1293]: Shouldn't at least the signature be unified ?

File: src/mono/mono/mini/mini-ops.h [hunk 2]
```diff
@@ -851,6 +851,7 @@ MINI_OP(OP_EXPAND_R8, "expand_r8", XREG, FREG, NONE)
 #if defined(TARGET_WASM)
 MINI_OP(OP_WASM_SIMD_BITMASK, "wasm_bitmask", IREG, XREG, NONE)
 MINI_OP3(OP_WASM_SIMD_SHUFFLE, "wasm_shuffle", VREG, XREG, XREG, XREG)
+MINI_OP(OP_WASM_SIMD_SUM, "wasm_sum", VREG, XREG, NONE)
 MINI_OP(OP_WASM_SIMD_SWIZZLE, "wasm_swizzle", VREG, XREG, XREG)
```
Comment (vargaz) [inline src/mono/mono/mini/mini-ops.h:855]: Is the return type really a VREG ?

Comment (radekdoulik) [reply src/mono/mono/mini/mini-ops.h]: Updated

File: src/mono/mono/mini/simd-intrinsics.c (hunk 1 above)
Comment (radekdoulik) [reply src/mono/mono/mini/simd-intrinsics.c:1293]: I used the same signature as amd64. I see that it changed meanwhile, so I will update it to use the same as arm64 and now amd64 as well.

**Review by vargaz — APPROVED** [review]

**Comment by radekdoulik** [comment]
> Is there a test for this in `src/tests`?

There are test for `Vector128` in `src\tests\JIT\HardwareIntrinsics\General\Vector128`. Do we run these for wasm?

**Pull request merged**